src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -1789,70 +1807,113 @@ public static int Min(this IEnumerable<int> source)
         public static long Min(this IEnumerable<long> source)
         {
             if (source == null) throw Error.ArgumentNull("source");
-            long value = 0;
-            bool hasValue = false;
-            foreach (long x in source)
+            long value;
+            using (IEnumerator<long> e = source.GetEnumerator())
             {
-                if (hasValue)
+                if (!e.MoveNext()) throw Error.NoElements();
+                value = e.Current;
+                while (e.MoveNext())
                 {
+                    long x = e.Current;
                     if (x < value) value = x;
                 }
-                else
-                {
-                    value = x;
-                    hasValue = true;
-                }
             }
-            if (hasValue) return value;
-            throw Error.NoElements();
+            return value;
         }
 
         public static long? Min(this IEnumerable<long?> source)
         {
             if (source == null) throw Error.ArgumentNull("source");
-            long? value = null;
-            foreach (long? x in source)
+            long? value;
+            using (IEnumerator<long?> e = source.GetEnumerator())
```

Comment:
My comments on the Min overloads for int and int? are applicable to the overloads for other types, but I'll refrain from making the same comments there. :)